### ISSUE TITLE
Win10/WSL2: Docker containers become unkillable

### ISSUE THREAD

**Issue by mason-hendricks** [docker/for-win#13320]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: `7297779D-1ECF-4C93-9816-379A68A6A548/20230320221923`

### Actual behavior
docker-compose down / Ctrl-C hangs until: `Error response from daemon: cannot stop container: {containerHash}: tried to kill container, but did not receive an exit event`.

Trying to docker exec into the container for a manual stop produces: `OCI runtime exec failed: exec failed: unable to start container process: error executing setns process: exit status 1: unknown`

Once this happens on my machine, all running containers are unstoppable until docker/wsl are stopped and restarted. I would prefer to not have to restart docker and each container when I just need one restarted.

Any guidance as to what could be causing this would be greatly appreciated!

### Expected behavior
docker-compose down or Ctrl-C in terminal should stop docker containers.

  - Windows Version: `10.0.19045`
  - Docker Desktop Version: `4.17.0`
  - WSL2 or Hyper-V backend?  `WSL2`
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: `No`

### Output of `& "C:\Program Files\Docker\Docker\resources\com.docker.diagnose.exe" check`

`C:\Program Files\Docker\Docker\resources>com.docker.diagnose.exe check
[2023-03-20T22:21:32.285664400Z][com.docker.diagnose.exe][I] set path configuration to OnHost
Starting diagnostics

[PASS] DD0027: is there available disk space on the host?
[PASS] DD0028: is there available VM disk space?
[PASS] DD0002: does the bootloader have virtualization enabled?
[SKIP] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0022: is the Virtual Machine Platform Windows Feature enabled?
[PASS] DD0021: is the WSL 2 Windows Feature enabled?
[PASS] DD0024: is WSL installed?
[PASS] DD0025: are WSL distros installed?
[PASS] DD0026: is the WSL LxssManager service running?
[PASS] DD0029: is the WSL 2 Linux filesystem corrupt?
[PASS] DD0035: is the VM time synchronized?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[PASS] DD0013: is the $PATH ok?
[PASS] DD0003: is the Docker CLI working?
[PASS] DD0005: is the user in the docker-users group?
[PASS] DD0038: is the connection to Docker working?
[PASS] DD0014: are the backend processes running?
[PASS] DD0007: is the backend responding?
[PASS] DD0008: is the native API responding?
[PASS] DD0009: is the vpnkit API responding?
[PASS] DD0010: is the Docker API proxy responding?
[PASS] DD0006: is the Docker Desktop Service responding?
[SKIP] DD0030: is the image access management authorized?
[PASS] DD0033: does the host have Internet access?
[PASS] DD0002: does the bootloader have virtualization enabled?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0022: is the Virtual Machine Platform Windows Feature enabled?
[PASS] DD0021: is the WSL 2 Windows Feature enabled?
[PASS] DD0024: is WSL installed?
[PASS] DD0025: are WSL distros installed?
[PASS] DD0026: is the WSL LxssManager service running?
[PASS] DD0029: is the WSL 2 Linux filesystem corrupt?
[PASS] DD0035: is the VM time synchronized?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[WARN] DD0032: do Docker networks overlap with host IPs? network {network} has subnet {network ip} which overlaps with host IP {host ip}

Please note the following 1 warning:

1 : The check: do Docker networks overlap with host IPs?
    Produced the following warning: network {network} has subnet {network ip} which overlaps with host IP {host ip}

If the subnet used by a Docker network overlaps with an IP used by the host, then containers
won't be able to contact the overlapping IP addresses.

Try configuring the IP address range used by networks: in your docker-compose.yml.
See https://docs.docker.com/compose/compose-file/compose-file-v2/#ipv4_address-ipv6_address

No fatal errors detected.`

**NOTE:** The above warning message is also produced on my laptop, which does not replicate this problem. Instead of switching machines, I'd prefer this to be rectified on my main work setup.

This doesn't seem to be limited to a docker configuration. My company uses 3 containers for our stack, any one of which can produce this issue.

**Comment by quantumtec** [comment]
I can confirm that using **Docker Desktop for Windows Version 4.17.1** with WSL can **breaks existing installations**:
Observed errors:
- some containers do not start
- downed containers block bound ports
- Error while Stopping containers (which currently should be starting, but never actually start)

We are getting reports from multiple affected machines. This is a serious regression. You should yank 4.17.1 and work on a fix asap. 
This seems to be related to the following changes:
- Improved debug information to better characterise failures under WSL 2.

**Comment by sam-thibault** [comment]
Which WSL kernel version are you running?  Did you update this recently?  There was a prior issue reported on Moby that identified the same error message for kernel version 5.15:
- https://github.com/moby/moby/issues/43094

**Comment by gustavozantut** [comment]
Same here, mainly when trying to use gpu , windows 11

**Comment by mason-hendricks** [comment]
Sorry for late reply, here's the kernel information 

`WSL version: 1.1.3.0
Kernel version: 5.15.90.1
WSLg version: 1.0.49
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.2728`

Could potentially be my issue, though this seemed to still be happening prior to my last update

**Comment by drie-cdts** [comment]
> I also noticed that when manually downloading the installer for version 4.17.1 from the [Docker homepage](https://docs.docker.com/desktop/release-notes/) the checksum reported on the page and the checksum I get are different.
> 
> ```
> Docker page checksum: e66a76f8f71f7e5e4488163d7da686d1fc669c5cbb47fec15085623c4deaf16a
> Checksum after download: 2EA284648A5F708428F3A06BB8E1EB68CBEBA6689B53C53D7CA24043A8F34800
> ```

Was about to file a new issue but found your comment. I can confirm that the checksums mismatch and I also get `2ea284648a5f708428f3a06bb8e1eb68cbeba6689b53c53d7ca24043a8f34800` when I generate the sha256 hash locally. Tested from two different machines with independent internet lines.

**Comment by mason-hendricks** [comment]
@sam-thibault Came back to this to say thanks and that I believe you found the fix for this. Newer WSL kernel versions seem to be the issue. I downgraded from 5.15 to 5.10 and as of today I haven't had any issues.

For anyone experiencing this problem, this fixed it for me.

- Remove your WSL package in elevated powershell with 
`$Package = Get-AppxPackage MicrosoftCorporationII.WindowsSubsystemforLinux -AllUsers 
Remove-AppxPackage $Package -AllUsers`

Follow 1-5 of this guide, don't do step 4.
https://learn.microsoft.com/en-us/windows/wsl/install-manual

Once installed, run a `wsl --update --inbox` Which excludes Microsoft store versions.
At this point I checked `wsl --status` and my kernel was 5.10, seems to have resolved the problem.
